### PR TITLE
SDI-50: ✨ Initial stab at netty / webflux

### PR DESCRIPTION
**Pull request by petergphillips**
- SDI-50: 🔧 Initial stab at netty / webflux
- SDI-50: 🔧 Initial stab at netty / webflux

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # HMPPS NOMIS Prisoner Mapping Service
 
 [![CircleCI](https://circleci.com/gh/ministryofjustice/hmpps-nomis-visits-mapping-service/tree/main.svg?style=svg)](https://circleci.com/gh/ministryofjustice/hmpps-nomis-visits-mapping-service)
-[![API docs](https://img.shields.io/badge/API_docs_-view-85EA2D.svg?logo=swagger)](https://nomis-visits-mapping-dev.hmpps.service.justice.gov.uk/swagger-ui/index.html?configUrl=/v3/api-docs)
+[![API docs](https://img.shields.io/badge/API_docs_-view-85EA2D.svg?logo=swagger)](https://nomis-visits-mapping-dev.hmpps.service.justice.gov.uk/webjars/swagger-ui/index.html?configUrl=/v3/api-docs)
```
Comment (petergphillips) [inline README.md:4]: not sure why, but using webflux version of spring docs changes the url

File: build.gradle.kts [hunk 2]
```diff
@@ -27,9 +28,8 @@ dependencies {
   runtimeOnly("org.springframework.boot:spring-boot-starter-jdbc")
   runtimeOnly("org.postgresql:postgresql:42.3.2")
 
-  implementation("org.springdoc:springdoc-openapi-ui:1.6.6")
+  implementation("org.springdoc:springdoc-openapi-webflux-ui:1.6.6")
   implementation("org.springdoc:springdoc-openapi-kotlin:1.6.6")
-  implementation("org.springdoc:springdoc-openapi-data-rest:1.6.6")
```
Comment (petergphillips) [inline build.gradle.kts:32]: taken out for now, was getting:
```
Consider defining a bean of type 'org.springframework.hateoas.server.LinkRelationProvider' in your configuration.
```

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -79,3 +79,7 @@ management:
       cache:
         time-to-live: 2000ms
 
+azure:
+  application-insights:
+    web:
+      enabled: false
```
Comment (petergphillips) [inline src/main/resources/application.yml:85]: needed otherwise pulls in servlet stuff

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/ClientTrackingWebFilter.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
+    val token = exchange.request.headers.getFirst(HttpHeaders.AUTHORIZATION)
+    if (token?.startsWith(bearer, ignoreCase = true) == true) {
+      try {
+        val jwtBody = getClaimsFromJWT(token)
+        val currentSpan = Span.current()
+        val user = jwtBody.getClaim("user_name")
+        user?.run { currentSpan.setAttribute(usernameKey, user.toString()) }
+        currentSpan.setAttribute(clientIdKey, jwtBody.getClaim("client_id").toString())
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/ClientTrackingWebFilter.kt:30]: note using opentelemetry to set attributes instead of app insights sdk.  This is as per the latest documentation in https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-in-process-agent#add-a-custom-dimension

Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/ClientTrackingWebFilter.kt:30]: can't get the test working at present since requires `Span` to be set, hence why test was removed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/AuthAwareTokenConverter.kt [hunk 5]
```diff
@@ -1,22 +1,22 @@
...
 import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
+import reactor.core.publisher.Mono
 
-class AuthAwareTokenConverter : Converter<Jwt, AbstractAuthenticationToken> {
+class AuthAwareTokenConverter : Converter<Jwt, Mono<AuthAwareAuthenticationToken>> {
   private val jwtGrantedAuthoritiesConverter: Converter<Jwt, Collection<GrantedAuthority>> =
     JwtGrantedAuthoritiesConverter()
 
-  override fun convert(jwt: Jwt): AbstractAuthenticationToken {
+  override fun convert(jwt: Jwt): Mono<AuthAwareAuthenticationToken> {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/AuthAwareTokenConverter.kt:15]: why did you need to do this, since I guess it wasn't blocking before?

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/AuthAwareTokenConverter.kt (hunk 5 above)
Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/AuthAwareTokenConverter.kt:15]: not sure what you mean - the contract on the class is `Converter<Jwt, Mono<AuthAwareAuthenticationToken>>` so have to return a `Mono`

Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/config/AuthAwareTokenConverter.kt:15]: But why is  contract `Converter<Jwt, Mono<AuthAwareAuthenticationToken>>`

and not 

`Converter<Jwt, AuthAwareAuthenticationToken>` 

?
